### PR TITLE
Fix `describe <table>` to work without SessionContext

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/7440

## Rationale for this change

I want to be able to run `DESCRIBE` in SQL without a `SessionContext`. 

## What changes are included in this PR?

1. Move the implementation of `DESCRIBE` planning to the physical planner
2. Add some tests

## Are these changes tested?
Yes

### CODE REVIEW

File: datafusion/core/src/execution/context.rs [hunk 1]
```diff
@@ -487,53 +481,6 @@ impl SessionContext {
         Ok(DataFrame::new(self.state(), plan))
     }
 
-    // return an record_batch which describe table
-    async fn return_describe_table_record_batch(
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:491]: This PR basically just moves this function into the physical planner

File: datafusion/core/tests/dataframe/describe.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+// under the License.
+
+use datafusion::{
+    assert_batches_eq,
+    prelude::{ParquetReadOptions, SessionContext},
+};
+use datafusion_common::{test_util::parquet_test_data, Result};
+
+#[tokio::test]
+async fn describe() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/tests/dataframe/describe.rs:25]: I just moved these tests into their own module to make it easier to find. Note that `Dataframe::describe` and the `DESCRIBE` sql command are different 😱

File: datafusion/core/tests/sql/describe.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        "| string_col      | Utf8                        | YES         |",
+        "| timestamp_col   | Timestamp(Nanosecond, None) | YES         |",
+        "| year            | Int32                       | YES         |",
+        "| month           | Int32                       | YES         |",
+        "+-----------------+-----------------------------+-------------+",
+    ];
+
+    assert_batches_eq!(expected, &results);
+
+    // also ensure we plan Describe via SessionState
```
Comment (alamb) [inline datafusion/core/tests/sql/describe.rs:51]: Without the code in this PR, this test fails like this:

```
---- sql::describe::describe_plan stdout ----
thread 'sql::describe::describe_plan' panicked at 'called `Result::unwrap()` on an `Err` value: Internal("Unsupported logical plan: DescribeTable must be root of the plan")', datafusion/core/tests/sql/describe.rs:55:38
stack backtrace:
   0: rust_begin_unwind
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:67:14
   2: core::result::unwrap_failed
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/result.rs:1651:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/result.rs:1076:23
   4: core_integration::sql::describe::describe_plan::{{closure}}
             at ./tests/sql/describe.rs:55:19
   5: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/future/future.rs:125:9
   6: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/future/future.rs:125:9
   7: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/current_thread/mod.rs:665:57
   8: tokio::runtime::coop::with_budget
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/coop.rs:107:5
   9: tokio::runtime::coop::budget
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/coop.rs:73:5
  10: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/current_thread/mod.rs:665:25
  11: tokio::runtime::scheduler::current_thread::Context::enter
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/current_thread/mod.rs:410:19
  12: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/current_thread/mod.rs:664:36
  13: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/current_thread/mod.rs:743:68
  14: tokio::runtime::context::scoped::Scoped<T>::set
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context/scoped.rs:40:9
  15: tokio::runtime::context::set_scheduler::{{closure}}
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context.rs:176:26
  16: std::thread::local::LocalKey<T>::try_with
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/thread/local.rs:270:16
  17: std::thread::local::LocalKey<T>::with
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/thread/local.rs:246:9
  18: tokio::runtime::context::set_scheduler
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context.rs:176:9
  19: tokio::runtime::scheduler::current_thread::CoreGuard::enter
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/current_thread/mod.rs:743:27
  20: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/current_thread/mod.rs:652:19
  21: tokio::runtime::scheduler::current_thread::CurrentThread::block_on::{{closure}}
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/current_thread/mod.rs:175:28
  22: tokio::runtime::context::runtime::enter_runtime
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context/runtime.rs:65:16
  23: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/current_thread/mod.rs:167:9
  24: tokio::runtime::runtime::Runtime::block_on
             at /Users/alamb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/runtime.rs:347:47
  25: core_integration::sql::describe::describe_plan
             at ./tests/sql/describe.rs:57:5
  26: core_integration::sql::describe::describe_plan::{{closure}}
             at ./tests/sql/describe.rs:24:27
  27: core::ops::function::FnOnce::call_once
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:250:5
  28: core::ops::function::FnOnce::call_once
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

File: datafusion/core/tests/dataframe/describe.rs (hunk 2 above)
Comment (comphead) [reply datafusion/core/tests/dataframe/describe.rs:25]: Dataframe `describe` is similar to pandas https://www.w3schools.com/python/pandas/ref_df_describe.asp

**Comment by comphead** [comment]
just a small neat on DESC without SessionContext: there is no plans to have a security layer in DF? because session context will help in this case, having some security token and allowing to know if the user can DESC table

**Comment by alamb** [comment]
>  just a small nit on DESC without SessionContext: there is no plans to have a security layer in DF? because session context will help in this case, having some security token and allowing to know if the user can DESC table

It is an interesting question about hooking up a security model to DataFusion -- at the moment I don't know of any plans, but having some sort of pluggable auth system would seem very much in keeping with DataFusion's long term goals.

I'll write up a ticket with some ideas over the next few days

**Comment by alamb** [comment]
Converting to draft until I sort out the errors and file a follow on ticket

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-datafusion/issues/7501 to track the security question

**Review by tustvold — APPROVED** [review]

**Pull request merged**